CRM/MembershipExtras/Hook/Pre/Membership.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+/**
+ * Implements hook to be run before a membership is created/edited.
+ */
+class CRM_MembershipExtras_Hook_Pre_Membership {
+
+  /**
+   * Parameters that will be used to create the membership.
+   *
+   * @var array
+   */
+  private $params;
+
+  /**
+   * Action being performed on the membership.
+   *
+   * @var string
+   */
+  private $op;
+
+  /**
+   * ID of the membership.
+   *
+   * @var int
+   */
+  private $id;
+
+  /**
+   * ID of the contribution that was created before the action was called for
+   * the membership.
+   *
+   * @var int
+   */
+  private $contributionID;
+
+  public function __construct($op, $id, &$params, $contributionID) {
+    $this->op = $op;
+    $this->id = $id;
+    $this->params = &$params;
+    $this->contributionID = $contributionID;
+  }
+
+  /**
+   * Preprocesses parameters used for Membership operations.
+   */
+  public function preProcess() {
+    switch ($this->op) {

```

Comment:
I don't think combinint the both in one class (CRM_MembershipExtras_Hook_Pre_Membership ) is the best course of action here, maybe just seperate them into : 

CRM_MembershipExtras_Hook_Pre_MembershipEdit 
CRM_MembershipExtras_Hook_Pre_MembershipCreate

if you did this then you can then git rid of the "now" confusing CRM_MembershipExtras_Hook_PreEdit_Membership by combing it into CRM_MembershipExtras_Hook_Pre_MembershipEdit